### PR TITLE
Adding PCRE2 tests to increase Regex code coverage

### LINKED ISSUE #61893: Port pcre2 regex tests
PCRE has lots of regex functional tests we should port over to improve our coverage and help maintain/validate compatibility:
https://github.com/PhilipHazel/pcre2/tree/master/testdata, e.g.
https://github.com/PhilipHazel/pcre2/blob/eb42305f072549facb3293f392f25768f35fa218/testdata/testoutput1

### PR DESCRIPTION
**Pull request by joperezr**
Fixes https://github.com/dotnet/runtime/issues/61893

This PR is adding all of the tests from the https://github.com/PCRE2Project/pcre2/testdata repo that can be parsed by our engine and don't rely on special syntax not supported by .NET (like `[[:digit:]]` syntax for example which we are able to parse, but it semantically means something different in our case)

I have already separately logged the issues I found in .NET while porting these (and most of them are already addressed by @stephentoub 😄). I also used Steve's technique for batching source-generator engine tests which helped bringing this single test execution from 2.3 minutes down to 1.1 seconds, which is why this test doesn't need to be moved to outerloop. In total, this is adding around 11K different tests into our test suite.

cc: @richlander since I'm updating one of our THIRD-PARTY-NOTICES file.

### CODE REVIEW

**Review by stephentoub — APPROVED** [review]
Excellent.

File: src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/RegexPcreTests.cs [hunk 1]
```diff
@@ -1552,7 +1552,10 @@ public static IEnumerable<(string Pattern, RegexOptions options, string Input, b
             yield return ("^\\p{Lo}", RegexOptions.None, "\u4D00", true);
             yield return ("^\\p{Lo}", RegexOptions.None, "\u4DB4", true);
             yield return ("^\\p{Lo}", RegexOptions.None, "\u4DB5", true);
-            yield return ("^\\p{Lo}", RegexOptions.None, "\u4DB6", true);
+            if (PlatformDetection.IsNotNetFramework)
+            {
+                yield return ("^\\p{Lo}", RegexOptions.None, "\u4DB6", true); // this was broken prior .NET 7
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/RegexPcreTests.cs:1557]: Might be clearer to say that it was fixed in .NET 7?

**Review by richlander — APPROVED** [review]

**Comment by danmoseley** [comment]
Nice!

**Comment by joperezr** [comment]
PR was green in the previous commit, and last one only addressed feedback on re-wording of comments, so I'll go ahead and merge now

**Pull request merged**